### PR TITLE
Explanation of Parsing vs. Compilation vs. Runtime

### PR DESCRIPTION
**Pull request by dbeatty10**
🎩 [Preview](https://github.com/dbt-labs/dbt-core/blob/dbeatty/parsing-compilation-etc/docs/eli64/parsing-vs-compilation-vs-runtime.md)

### Problem

> There’s a lot of confusion about what dbt does at parse time vs. compile time / runtime. Even that separation is a relative simplification: parsing includes multiple steps, and while there are some distinctions between "compiling" and "running" a model, the two are **very** closely related.
> 
> It's come up many times before, and we expect it will keep coming up! A decent number of bug reports in `dbt-core` are actually rooted in a misunderstanding of when configs are resolved, especially when folks are using pre/post hooks, or configs that alter materialization behavior (`partitions`, `merge_exclude_columns`, etc).

### Solution

So let's publish something that we can use as a reference! Markdown within the code base would be a good first step.

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me  
- [x] ~I have run this code in development and it appears to resolve the stated issue~
- [x] Tests are not required/relevant for this PR
- [x] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc)
- [x] ~This PR includes [type annotations](https://docs.python.org/3/library/typing.html) for new and modified functions~

### CODE REVIEW

File: core/dbt/parser/parsing-vs-compilation-vs-runtime.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+- **Jinja-SQL** wants to compile down to "vanilla" SQL, appropriate for this database, where any calls to `ref('something')` have been replaced with `database.schema.something`.
+- dbt doesn’t directly modify or rewrite user-provided **Python** code at all. Instead, "compilation" looks like code generation: appending more methods that allow calls to `dbt.ref()`, `dbt.source()`, and `dbt.config.get()` to return the correct results at runtime.
+
+</details>
+
+<details>
+<summary>If your model’s code uses a dynamic query to template code, this requires a database connection.</summary>
+
+- At this point, `[execute](https://docs.getdbt.com/reference/dbt-jinja-functions/execute)` is set to `True`.
```
Comment (jtcohen6) [inline core/dbt/parser/parsing-vs-compilation-vs-runtime.md:89]:
```suggestion
- At this point, <a href="https://docs.getdbt.com/reference/dbt-jinja-functions/execute">execute</a>` is set to `True`.
```

File: core/dbt/parser/parsing-vs-compilation-vs-runtime.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+### Two potential optimizations
+
+1. [**"Partial parsing."**](https://docs.getdbt.com/reference/parsing#partial-parsing) dbt saves the mostly-done Manifest from last time, in a file called `target/partial_parse.msgpack`. dbt **just** reads the files that have changed (based on file system metadata), and makes partial updates to that mostly-done Manifest. Of course, if a user has updated configuration that could be relevant globally (e.g. `dbt_project.yml`, `--vars`), we have to opt for a full re-parse — better safe (slow & correct) than sorry (fast & incorrect).
+2. **A dbt Server.** Both the new `dbt-server` and the old `dbt-rpc` server have mechanisms to separate parsing from execution. They save a Manifest for reuse between commands. When files change, they trigger a re-parse and re-construction of the Manifest behind the scenes. This way, you don’t have to wait for re-parsing (even partial parsing) when you actually submit a command; it’s ready for you. The new approach, taken by `dbt-server` + Runtime, is much more resilient and less brittle than the previous approach in `dbt-rpc`.
```
Comment (jtcohen6) [inline core/dbt/parser/parsing-vs-compilation-vs-runtime.md:149]: Let's update (2) to be about "manifest reuse" instead of dbt-rpc / dbt Server. Something like:
```suggestion
2. Reusing manifests: https://docs.getdbt.com/reference/programmatic-invocations#reusing-objects. Note that this is taking "full control," and there are failure modes (example: [dbt-core#7945](https://github.com/dbt-labs/dbt-core/issues/7945)).
```

File: core/dbt/parser/parsing-vs-compilation-vs-runtime.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+## Appendix
+
+<details>
+<summary>Click to toggle notes on parsing</summary>
+
+### Notes on parsing
+
+- **dbt has not yet connected to a database.** Every step performed thus far has required only project files, configuration, and `dbt-core`. You can perform parsing without an Internet connection.
+- There is a command called `parse`, which does **just** "parsing" + "resolving," as a way to measure parsing performance in large projects. That command *could* write `manifest.json` once it's done; it doesn't, today, for no particular reason. (We're thinking about changing this.)
```
Comment (jtcohen6) [inline core/dbt/parser/parsing-vs-compilation-vs-runtime.md:143]:
```suggestion
- There is a command called `parse`, which does **just** "parsing" + "resolving," as a way to measure parsing performance in large projects. That command is the fastest way to write `manifest.json` (since v1.5).
```

File: core/dbt/parser/parsing-vs-compilation-vs-runtime.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+The outcome of compiling a model is updating its Manifest entry in two important ways:
+- `compiled` is set to `True`
+- `compiled_code` is populated with (what else) the compiled code for this model
+
+### Running / materializing a model
+
+A model’s `compiled_code` is passed into the materialization macro, and the materialization macro is executed. That materialization macro will also call user-provided pre- and post-hooks, and other built-in macros that return the appropriate DDL + DML statements (`create`, `alter`, `merge`, etc.)
+
+(It’s set to a context variable named [`sql`](https://github.com/dbt-labs/dbt-core/blob/16f529e1d4e067bdbb6a659a622bead442f24b4e/core/dbt/context/providers.py#L1314-L1323), which you’ll see in some materializations — and we should really change that to just `model['compiled_code']`.)
```
Comment (jtcohen6) [inline core/dbt/parser/parsing-vs-compilation-vs-runtime.md:122]:
```suggestion
(For legacy reasons, `compiled_code` is also available as a context variable named [`sql`](https://github.com/dbt-labs/dbt-core/blob/16f529e1d4e067bdbb6a659a622bead442f24b4e/core/dbt/context/providers.py#L1314-L1323). You'll see it referenced as `sql` in some materializations. Going forward, `model['compiled_code']` is a better way to access this.)
```

File: core/dbt/parser/parsing-vs-compilation-vs-runtime.md [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+- The code for this is *gnarly*. That’s all I’m going to say about it for now.
+
+</details>
+
+<details>
+<summary>When compiling happens for a given node varies by command.</summary>
+
+- In `dbt compile`, every model is compiled concurrently, up to the number of threads, rather than in DAG order.
+    - For example, if one model’s templated SQL depends on an introspective query that expects another model to have already been materialized, this can lead to errors.
+    - `dbt compile` has not historically supported `--defer`, but this was added in v1.3 (with [one known bug](https://github.com/dbt-labs/dbt-core/issues/6124)).
```
Comment (jtcohen6) [inline core/dbt/parser/parsing-vs-compilation-vs-runtime.md:107]: Let's cut this bullet - irrelevant for most recent versions

File: core/dbt/parser/parsing-vs-compilation-vs-runtime.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+<summary>Compilation is also when ephemeral model CTEs are interpolated into the models that `ref` them.</summary>
+
+- The code for this is *gnarly*. That’s all I’m going to say about it for now.
+
+</details>
+
+<details>
+<summary>When compiling happens for a given node varies by command.</summary>
+
+- In `dbt compile`, every model is compiled concurrently, up to the number of threads, rather than in DAG order.
```
Comment (jtcohen6) [inline core/dbt/parser/parsing-vs-compilation-vs-runtime.md:105]: Since I wrote this, I figured out that it isn't actually how it works today. It _should_ work this way:
- https://github.com/dbt-labs/dbt-core/issues/7914

Let's cut the bullet for now

File: core/dbt/parser/parsing-vs-compilation-vs-runtime.md [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+<details>
+<summary>(Toggle for many more details.)</summary>
+
+- (Because your SQL and YAML live in separate files, this is actually two steps. But for things like `sources`, `exposures`, `metrics`, `tests`, it’s a single pass.)
+- dbt needs to capture and store two vital pieces of information: **dependencies** and **configuration**.
+    - We need to know the shape of the DAG. That includes which models are disabled, in addition to which models will be depending on which other models.
+    - Plus, certain configurations have implications for **node selection**, which supports selecting models using the `tag:` and `config:` methods.
+- Parsing also resolves the configuration for that model, based on configs set in `dbt_project.yml`, and macros like `generate_schema_name`. (These are "special" macros, whose results are saved at parse time!)
+- The way dbt parses models depends on the language that model is written in.
+    - Python models are statically analyzed using the Python AST. [This is pretty fast.](https://www.notion.so/f113894b4cb7412e83b7f9fdb3a24cdb?pvs=21)
```
Comment (jtcohen6) [inline core/dbt/parser/parsing-vs-compilation-vs-runtime.md:30]: This is a link to an internal Notion doc. We should either remove, or qualify the "pretty fast" by picking a number from the doc

**Review by jtcohen6 — COMMENTED** [review]
Thanks for making this happen @dbeatty10! Talk is cheap, PRs are precious ;)

- I left a first round of comments, mostly catching things that have changed since I initially wrote this last December
- As far as where in the repo this should land - I wonder if the [`docs/architecture`](https://github.com/dbt-labs/dbt-core/tree/main/docs) subfolder could be the right place? Perhaps (in the future) accompanied by some diagrams?

File: core/dbt/parser/parsing-vs-compilation-vs-runtime.md [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+## What is "parsing"?
+
+**In a sentence:** dbt reads all the files in your project, and constructs an internal representation of the project ("manifest").
+
+To keep it really simple, let’s say this happens in two steps: "Parsing" and "Resolving."
+
+### Parsing
+
+As a user, you write models as SQL + YAML. dbt wants to understand that model as a Python object, defined by an internal data structure. It also wants to know its dependencies and configuration (= its place in the DAG). dbt reads your code **for that one model,** and attempts to construct that object, raising a **validation** error if it can’t.
```
Comment (MichelleArk) [inline core/dbt/parser/parsing-vs-compilation-vs-runtime.md:19]: nit, feel free to disregard: I think switching 'that model' with 'each model' would read more clearly in this sentence relative to the plural 'models' in the beginning

File: core/dbt/parser/parsing-vs-compilation-vs-runtime.md [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+<details>
+<summary>(Toggle for many more details.)</summary>
+
+- (Because your SQL and YAML live in separate files, this is actually two steps. But for things like `sources`, `exposures`, `metrics`, `tests`, it’s a single pass.)
+- dbt needs to capture and store two vital pieces of information: **dependencies** and **configuration**.
+    - We need to know the shape of the DAG. That includes which models are disabled, in addition to which models will be depending on which other models.
```
Comment (MichelleArk) [inline core/dbt/parser/parsing-vs-compilation-vs-runtime.md:26]: is there something missing after 'will be...'?

File: core/dbt/parser/parsing-vs-compilation-vs-runtime.md (hunk 8 above)
Comment (dbeatty10) [inline core/dbt/parser/parsing-vs-compilation-vs-runtime.md:19]:
```suggestion
As a user, you write models as SQL + YAML. dbt wants to understand each model as a Python object, defined by an internal data structure. It also wants to know its dependencies and configuration (= its place in the DAG). dbt reads your code **for that one model,** and attempts to construct that object, raising a **validation** error if it can’t.
```

File: core/dbt/parser/parsing-vs-compilation-vs-runtime.md [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+### Parsing
+
+As a user, you write models as SQL + YAML. dbt wants to understand each model as a Python object, defined by an internal data structure. It also wants to know its dependencies and configuration (= its place in the DAG). dbt reads your code **for that one model,** and attempts to construct that object, raising a **validation** error if it can’t.
+
+<details>
+<summary>(Toggle for many more details.)</summary>
+
+- (Because your SQL and YAML live in separate files, this is actually two steps. But for things like `sources`, `exposures`, `metrics`, `tests`, it’s a single pass.)
+- dbt needs to capture and store two vital pieces of information: **dependencies** and **configuration**.
+    - We need to know the shape of the DAG. That includes which models are disabled, in addition to which models will be depending on which other models.
```
Comment (dbeatty10) [inline core/dbt/parser/parsing-vs-compilation-vs-runtime.md:26]: I'm guessing this was the intent? Rolling with it.
```suggestion
    - We need to know the shape of the DAG. This includes which models are disabled. It also includes dependency relationships between models.
```

File: core/dbt/parser/parsing-vs-compilation-vs-runtime.md (hunk 7 above)
Comment (dbeatty10) [reply core/dbt/parser/parsing-vs-compilation-vs-runtime.md:30]: Removing

File: core/dbt/parser/parsing-vs-compilation-vs-runtime.md [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+<details>
+<summary>(Toggle for many more details.)</summary>
+
+- (Because your SQL and YAML live in separate files, this is actually two steps. But for things like `sources`, `exposures`, `metrics`, `tests`, it’s a single pass.)
+- dbt needs to capture and store two vital pieces of information: **dependencies** and **configuration**.
+    - We need to know the shape of the DAG. This includes which models are disabled. It also includes dependency relationships between models.
+    - Plus, certain configurations have implications for **node selection**, which supports selecting models using the `tag:` and `config:` methods.
+- Parsing also resolves the configuration for that model, based on configs set in `dbt_project.yml`, and macros like `generate_schema_name`. (These are "special" macros, whose results are saved at parse time!)
+- The way dbt parses models depends on the language that model is written in.
+    - Python models are statically analyzed using the Python AST. [This is pretty fast.](https://www.notion.so/f113894b4cb7412e83b7f9fdb3a24cdb?pvs=21)
```
Comment (dbeatty10) [inline core/dbt/parser/parsing-vs-compilation-vs-runtime.md:30]:
```suggestion
    - Python models are statically analyzed using the Python AST.
```

File: core/dbt/parser/parsing-vs-compilation-vs-runtime.md [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+
+- **Jinja-SQL** wants to compile down to "vanilla" SQL, appropriate for this database, where any calls to `ref('something')` have been replaced with `database.schema.something`.
+- dbt doesn’t directly modify or rewrite user-provided **Python** code at all. Instead, "compilation" looks like code generation: appending more methods that allow calls to `dbt.ref()`, `dbt.source()`, and `dbt.config.get()` to return the correct results at runtime.
+
+</details>
+
+<details>
+<summary>If your model’s code uses a dynamic query to template code, this requires a database connection.</summary>
+
+- At this point, `[execute](https://docs.getdbt.com/reference/dbt-jinja-functions/execute)` is set to `True`.
```
Comment (dbeatty10) [inline core/dbt/parser/parsing-vs-compilation-vs-runtime.md:89]:
```suggestion
- At this point, [`execute`](https://docs.getdbt.com/reference/dbt-jinja-functions/execute) is set to `True`.
```

File: core/dbt/parser/parsing-vs-compilation-vs-runtime.md [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+<summary>Compilation is also when ephemeral model CTEs are interpolated into the models that `ref` them.</summary>
+
+- The code for this is *gnarly*. That’s all I’m going to say about it for now.
+
+</details>
+
+<details>
+<summary>When compiling happens for a given node varies by command.</summary>
+
+- In `dbt compile`, every model is compiled concurrently, up to the number of threads, rather than in DAG order.
```
Comment (dbeatty10) [inline core/dbt/parser/parsing-vs-compilation-vs-runtime.md:105]:
```suggestion
```

File: core/dbt/parser/parsing-vs-compilation-vs-runtime.md [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+
+- The code for this is *gnarly*. That’s all I’m going to say about it for now.
+
+</details>
+
+<details>
+<summary>When compiling happens for a given node varies by command.</summary>
+
+    - For example, if one model’s templated SQL depends on an introspective query that expects another model to have already been materialized, this can lead to errors.
+    - `dbt compile` has not historically supported `--defer`, but this was added in v1.3 (with [one known bug](https://github.com/dbt-labs/dbt-core/issues/6124)).
```
Comment (dbeatty10) [inline core/dbt/parser/parsing-vs-compilation-vs-runtime.md:106]:
```suggestion
```

File: core/dbt/parser/parsing-vs-compilation-vs-runtime.md [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+- In large projects, the parsing step can also be quite slow: reading lots of files, doing lots of dataclass validation, creating lots of links between lots of nodes. (See below for details on two potential optimizations.)
+
+### Two potential optimizations
+
+1. [**"Partial parsing."**](https://docs.getdbt.com/reference/parsing#partial-parsing) dbt saves the mostly-done Manifest from last time, in a file called `target/partial_parse.msgpack`. dbt **just** reads the files that have changed (based on file system metadata), and makes partial updates to that mostly-done Manifest. Of course, if a user has updated configuration that could be relevant globally (e.g. `dbt_project.yml`, `--vars`), we have to opt for a full re-parse — better safe (slow & correct) than sorry (fast & incorrect).
+2. Reusing manifests: https://docs.getdbt.com/reference/programmatic-invocations#reusing-objects. Note that this is taking "full control," and there are failure modes (example: [dbt-core#7945](https://github.com/dbt-labs/dbt-core/issues/7945)).
```
Comment (dbeatty10) [inline core/dbt/parser/parsing-vs-compilation-vs-runtime.md:147]:
```suggestion
2. [**"Reusing manifests."**](https://docs.getdbt.com/reference/programmatic-invocations#reusing-objects) Note that this is taking "full control," and there are failure modes (example: [dbt-core#7945](https://github.com/dbt-labs/dbt-core/issues/7945)).
```

File: core/dbt/parser/parsing-vs-compilation-vs-runtime.md [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+<summary>Compilation is also when ephemeral model CTEs are interpolated into the models that `ref` them.</summary>
+
+- The code for this is *gnarly*. That’s all I’m going to say about it for now.
+
+</details>
+
+<details>
+<summary>When compiling happens for a given node varies by command.</summary>
+
+    - For example, if one model’s templated SQL depends on an introspective query that expects another model to have already been materialized, this can lead to errors.
```
Comment (dbeatty10) [inline core/dbt/parser/parsing-vs-compilation-vs-runtime.md:105]:
```suggestion
- For example, if one model’s templated SQL depends on an introspective query that expects another model to have already been materialized, this can lead to errors.
```

**Comment by dbeatty10** [comment]
> * As far as where in the repo this should land - I wonder if the [`docs/architecture`](https://github.com/dbt-labs/dbt-core/tree/main/docs) subfolder could be the right place? Perhaps (in the future) accompanied by some diagrams?

I'm guessing we'll want to move and/or rename this at least one more time. But in the interim, moved it here:
`docs/eli64/parsing-vs-compilation-vs-runtime.md`

> when ELI5 just isn't detailed enough

File: docs/eli64/parsing-vs-compilation-vs-runtime.md [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+
+## What is "parsing"?
+
+**In a sentence:** dbt reads all the files in your project, and constructs an internal representation of the project ("manifest").
+
+To keep it really simple, let’s say this happens in two steps: "Parsing" and "Resolving."
+
+### Parsing
+
+As a user, you write models as SQL + YAML. dbt wants to understand each model as a Python object, defined by an internal data structure. It also wants to know its dependencies and configuration (= its place in the DAG). dbt reads your code **for that one model,** and attempts to construct that object, raising a **validation** error if it can’t.
```
Comment (MichelleArk) [inline docs/eli64/parsing-vs-compilation-vs-runtime.md:19]: SQL (or Python!) + YAML

Comment (dbeatty10) [reply docs/eli64/parsing-vs-compilation-vs-runtime.md]: Good call!

Attempted to address this in https://github.com/dbt-labs/dbt-core/pull/8744/commits/4fc26eb82d9e9c80475b400cd3480249db1df65f
